### ISSUE TITLE
Redesign of UBDisplayManager

### ISSUE THREAD

**Issue by letsfindaway** [letsfindaway/OpenBoard#74]
The `UBDisplayManager` needs a redesign due to several reasons:

- it uses the obsolete `QApplication::desktop()` function and the related `QDesktopWidget` class. This needs to be changed at the latest when Qt 6 porting becomes relevant.
- It is not flexible in handling setups with more than 2 screens. There are several user requests to improve this.
- It should be prepared to support HiDPI displays.

Finally, a redesign should encapsulate as much of the screen management as possible. It shall no longer be necessary that other classes care about screen assignment any longer and need to handle screens and global coordinates. Instead this all should be covered by a redesigned `UBDisplayManager`.

The following is a list of things I have already discovered. This list will certainly be extended.

- [x] Move ownership of the `UBDisplayManager` to `UBApplication` and initialize it early, so that all other component can use it.
- [x] Add methods to retrieve size and physical and logical DPI of screens by screen role.
- [x] Take care to distinguish between physical and logical DPI values to better support HiDPI monitors. Currently only physicalDpi is used, even if logicalDpi would be the right choice.
- [x] Add methods to grab screen content.
- [x] Remove methods exposing `QScreen` objects or global coordinates.
- [x] Substitute each other use of `QApplication::desktop()` or `QDesktopWidget` by invoking a method of the `UBDisplayManager`.
- [x] Add method to help filling the model of a `QComboBox` for selecting a screen, e.g. in the preferences dialog, and to assign a screen to a role.

This redesign will be developed on top of the `1.7-dev` branch as it is relevant for that upcoming version. Everything developed within this branch will be compatible with Qt 6.

**Comment by letsfindaway** [comment]
# DPI values

One usage of `QApplication::desktop()` is to retrieve the DPI of the screens. Currently, `physicalDPI` of the complete desktop is used everywhere and there is no awareness that different screens could have different device pixel ratio. The redesign of the `UBDisplayManager` should also take care of that to support HiDPI monitors. See https://doc.qt.io/qt-5/highdpi.html for more information.

Short summary:

- When using pixels to draw or position a widget, to paint a line, etc. device independent pixels are used (so.called user space).
- These are also used to define the size of a font.
- This corresponds to the logical DPI of the monitor.
- This coordinate system is also used when asking for `QScreen::size()`.
- When drawing a pixmap, the `devicePixelRatio` of the pixmap is relevant.

**Comment by letsfindaway** [comment]
## UBExportFullPDF

When analyzing the usage of the DPI value I found the following:

This class has a member variable `mHasPDFBackgrounds` which is set to `false` in line 107 and never set to any other value anywhere. Some lines later in line 111 if this variable is `false`, then just a rename is executed. The complete `else` branch from line 116 to 216 is never executed and can be deleted.

All usages of the scale factor are within this obsolete `else` branch. So the complete variable `mScaleFactor` can be removed.

**Comment by letsfindaway** [comment]
## UBExportPDF

This class now does the work! It uses a DPI value to calculate a `scaleFactor` expressing the "points per pixel" ratio. This is then later used to convert a page size in pixels in a `QPageSize` in points.

The original page size is in logical pixels, the same as widget sizes etc are expressed. We should therefore also use a logical DPI value. To be explicit, we should use the logical DPI value of the control screen.

**Comment by letsfindaway** [comment]
## UBImportPDF

This is more complicated. It is used to instruct the PDF renderer to create an image of the desired pixel density. If we want to create a HiDPI render image, then we would probably have to use the physical DPI value and set the proper device pixel ratio in the rendered image. If not, then the logical DPI would be ok.

I think for full HiDPI support this information would be helpful. For that the `PDFRenderer` would need a new function to set the device pixel ratio. For now (the redesign of the `UBDisplayManager`) this is not needed, but the necessary information would be available if it gets important at some later time.

**Comment by letsfindaway** [comment]
## UBCCFSubsetAdaptor and UBSvgSubsetAdaptor

Here the use of the `QDisplayWidget` can be replaced by the appropriate call to the `UBDisplayManager`. In all places the `logicalDpi` should be used.

**Comment by letsfindaway** [comment]
## UBBoardController

The comments in `UBBoardController::initBackgroundGridSize()` are no longer true: attaching the `UBDisplayManager` to `UBApplication` also initializes it much earlier, so that screen configuration is already known at this point in time.

**Comment by letsfindaway** [comment]
# Screen count

Another use of `QDesktopWidget` is the screen count and detecting addition and removal of screens. This should completely be handled by the redesigned `UBDisplayManager`, which should also offer an appropriate signal if the screen count changes.

Note that there is also that check for display with same geometry, which adjusts the physical screen count to a number of screens usable for a specific role. This must be reflected in that API. This calculation should not only be triggered when a screen is added or removed, but also when the screen geometry changes, as this may indicate that the result of this adjustment changes.

**Comment by letsfindaway** [comment]
## Screen grabbing

The `QDesktopWidget` is also used to get the effenctive window id used to grab screen content with `QScreen->grabWindow`. I think the idea is to use the `QWidget::effectiveWinId()` function on the `QDesktopWidget? in order to get a window id which can be used to grab the complete screen.

This seems however not to be necessary. There is a [Qt 6 screenshot example](https://doc.qt.io/qt-6.2/qtwidgets-desktop-screenshot-example.html) which is just using the constant `0` for this purpose! Tests showed that this works.

### UBDesktopAnnotationController

Grabs the desktop content for screenshots in `getScreenPixmap()`.

**Comment by letsfindaway** [comment]
## Screen Configuration

Currently screen configuration is based on the sequence how `QScreen` objects are returned by Qt, the notion of the "primary display" and two check boxes:

- By default the primary screen is the control display and the next in list is the display screen.
- All others are assigned to the "Previous" displays.
- The Control and Display screen can be switched.
- The multi-monitor mode can be switched-off.

With this information the `UBDisplayManager` assigns roles to the screens. 

To be close to this approach, but still flexible I propose just to replace the flag for switching Control and Display by a configurable list of screen names to be used. In the configuration menu:
![grafik](https://user-images.githubusercontent.com/35591413/168220469-654cd00f-adf1-4628-887a-e5398b771416.png)
the second line would be replaced by a line edit field containing this list.

To help filling out this list, it will provide a special mode:

- When putting the focus on this field, the screen names will be displayed on each screen.
- Pressing `Backspace` or probably by some "Clear all" button the field content can be deleted one by one entry or completely.
- By clicking on the screen names on the individual screens, this screen will be added to the list. The screen name will then be greyed out and no longer be available for selection.

In this way you can easily create a list of used screen names. OpenBoard will then use the intersection of this list and the actually available screens. Still the boolean "Use multiple screens" will be available to restrict OpenBoard to a single screen (the first in the list).

**Comment by kaamui** [comment]
I tried to make more tests to find a finer reproduction scenario. I unplugged my wacom and relaunched OpenBoard. When going to the preferences, my input logically still has "DP-2-1,DP-1,DP-2-2" as I let them, but backspace key is inoperative. I can use the delete icon though. If I close the preferences, the input I deleted comes back. If I delete it then enter a proper setup ("DP-1, DP-2-2"), screens are properly placed, and everything works fine again in terme of screen placement.

**Comment by letsfindaway** [comment]
> I tried to make more tests to find a finer reproduction scenario. I unplugged my wacom and relaunched OpenBoard. When going to the preferences, my input logically still has "DP-2-1,DP-1,DP-2-2" as I let it, but backspace key is inoperative. I can use the delete icon though. If I close the preferences, the input I deleted comes back. If I delete it then enter a proper setup ("DP-1, DP-2-2"), screens are properly placed, and everything works fine again in terms of screen placement.

I can at least explain that, but it needs some fixing.

- The stored configuration stays the same, even when one of the devices is currently unavailable. This intended.
- The backspace key does not work, because then the validator jumps in and detects that the input can not completed to a valid list. This needs to be fixed, backspace must also work in this scenario!
- If you delete everything, then this is considered an invalid configuration and is therefore not persisted. Only valid input is persisted. Therefore the old list comes back when reopening the preferences. Do you think this should be changed? How? Sould we probably consider an empty list as "valid" and interpret it as "use the primary monitor only"?
- When you delete and enter a proper setup, then this is again a "valid" configuration and persisted.

**Comment by kaamui** [comment]
Ok, that's what I thought. I think users can be very frustrated if they don't understand why typing is not working. Maybe one simple solution (if you can stop them from typing wrong names then you know the exact available list) is to display in a small text, below the input, the list of available names. 

But as I stated in the "Display Names" issue, I wonder if we shouldn't use a simpler (not simpler to implement) naming convention like it is done in Ubuntu (with 1-based indexes and the rectangles you make appear it would be really simple and user-friendly).

If so, displaying a text like "Available screens are : 1, 2, 3" might look a bit strange, I don't know... now that I wrote it I find it less stranger than I thought

I tend to prefer letting the user doing everything he wants, but providing a safe environment so he can't hurt himself ^^. Maybe a simple red border/background when input is not vaild, with the consequence that input will not be persisted (like using an already taken name in the Documents Mode).

Empty input could indeed be considered as valid. I think it should be interpreted as "use the primary monitor only" in single screen mode, and "use them all from left to right" in multi screen mode. We should add an apply button right after the input, I think changes on the fly is a bit too aggressive in terms of what's happening on the screens :D  

> For the window mode problems you reported above I need my development computer which I do not have here at the openSUSE conference. I will look into this next week at the latest.

I'll try to look at it soon too, when 1.6.2 finally comes out (in schools). Thank you, as usual, for your help ! ^^

**Comment by letsfindaway** [comment]
Trying to understand what happens here, I could actually not reproduce that on my setup.
> ##### Ubuntu
> 
> * `runInWindow` seems to no longer work at OpenBoard's launch (always in full screen mode (visually at least))

Here the difference is that in window mode there is a window title bar and the window can be resized with the mouse. The window however occupies the full screen. Maybe that with other window managers the title bar is outside the visible area. So it might be better to add some margin around the window when starting in window mode. Currently I indeed make the window as large as the screen.

> * When clicking on the input to change display configuration, human-readable display names (rectangles) appear and blink until I click a second time on the input. Same if I click on the delete icon

I cannot observe that. Note that the appearance of the labels is attached to the focus being on the input field. For me it seems that it takes several clicks on the input field to reliably put the focus on that, and that e.g. making the labels visible takes the focus away from the input field. I will try to explicitly reassign the focus to the input field after displaying the labels.

> * `runInWindow=false`
>       
>   * single screen --> multi screen
>         
>     * control screen and display screen are on the same screen (wacom). The third screen ("Previous Page" screen) is always on the good position.

Cannot reproduce that. Screens were always where expected for me.

> * `runInWindow=true`
>       
>   * single screen --> multi screen
>         
>     * control screen is in windowed mode (but occupies a little portion of the screen) and is behind the display screen (both on the same screen)
>   * multi screen --> single screen
>         
>     * control screen is in windowed mode (but occupies a little portion of the screen, even after i tried to maximize it, exit, relaunch, and switch again single/multi screen)

Cannot reproduce that either. For me windows are never behind each other. Just an idea; I already have some "special" code at `UBDisplayManager::addOrRemoveScreen`: Here I delay the processing by quite some time, because the window manager also tries to position windows and I have to re-do that afterwards. Probably in your configuration there is something similar? The small windows you experience are quite likely not sized by OpenBoard, but by the window manager. So their position is probably also set by the window manager and not by OpenBoard. 

> * when changing display configuration while `runInWindow=true`, control screen returns to fullScreen mode

As each change of configuration is a full reconfiguration of screens and widgets, it gets again the full screen size. This should be avoided.

**Comment by kaamui** [comment]
Ok so now it is quite pretty sure to me that it has nothing to do with resolution, or screens placement, just with Gnome Dock. 

I just tried with the Gnome Dock place on the third screen, and now when I try to place the second display on it, it goes to the second screen : 
![Capture d’écran de 2022-06-07 13-53-02](https://user-images.githubusercontent.com/11720903/172373035-6601ae40-8afa-48b7-89f8-e8f787eff867.png)

**Comment by letsfindaway** [comment]
Hi @kaamui,
I could now also create some scenarios where the wrong screens are used. For me it seems that `setGeometry()` is quite picky about whether it does what is intended or not. Will have to try that more systematically. If you have interference with the Gnome dock, then it seems to me that window managers have their own opinion about window placement and in some cases revert it as they like. I will make some experiments with deferred calls to `setGeomentry()` in order to overwrite what the window manager means.

**Comment by letsfindaway** [comment]
## Collection of work items

In this comment I will collect and comment all identified issues and work items. It will grow over time and shall provide an overview of what is done and needs to be done.

### Explicitly apply screen configuration changes

Currently changes of the screen configuration are applied on-the-fly as soon as a valid entry is provided in the screen list. There is the idea to replace this by explicit application of changes. I think there are some points to consider:

- All other configuration changes in the preferences dialog are applied on-the-fly, also the previous checkbox for switching control and display screen. Why should it be different for screen configuration?
- Where should the "Apply" button be located? The first page of the preferences dialog is already quite full.

### Popup menu issue

The following issue appearing on Windows has to be verified and analyzed:

> in FullScreen Mode, After going in Web Mode, main menu and context menus don't popup anymore. Returning to Board Mode, I open a textBox, right click on it and same issue. I think there is an issue with the window properties state (another clue : on my own PC at home, when entering Internal web browser, GeForce Experience overlay shows up like if I was entering in a game (fullscreen). I tried to record it to show you but recording stops when clicking on Web Mode button (maybe related to changes in the window properties ?)

### Screen blinking issue

The following issue appearing on Windows has to be verified and analyzed:

> when switching between the different modes (Desktop,Web, Board...), there's some systematic blinking on every screen (3-4 times), before it stabilizes

### Focus issue

The following issue appearing on Ubuntu seems to be related with loosing and gaining focus.

> When clicking on the input to change display configuration, human-readable display names (rectangles) appear and blink until I click a second time on the input. Same if I click on the delete icon

Check whether this can be resolved by increasing the screen label fade-out timer.

### Use screen index instead of screen name in the UI

The idea behind this point is to decouple from whatever screen names a specific operating system or window manager uses and to create a simple, not-surprising UI. While I support this point, there are still some questions open:

- How does the mapping to screen indexes work?
- Will the mapping change depending on which monitors are connected? If yes, is this then not again surprising to the user?
- What happens when a screen is connected or disconnected while the configuration dialog is open? Will we have to recalculate the mapping?

### Allow empty list

If the list of screens is empty, this could be interpreted like the original multi-screen mode: Take and use all screens in the order as they are delivered from the OS. This would require that we distinguish between

- The configuration entry does not exist at all. This is the case when we upgrade from a previous version of OpenBoard.
- The configuration entry exists, but is empty. This is then the explicit configuration of the described multi-screen mode. To indicate this to the user we could then display a placeholder text in the line edit field stating "Use all connected screens".

### Choose proper timer values

We currently have timers for the following purposes:

- Fade-out of the screen labels after the line edit field looses focus.<br>This is used to avoid that the screen labels disappear when the line edit field shortly looses focus, e.g. because somebody clicked on one of the labels. This timer is currently 250ms.
- Screen reconfiguration after connecting or disconnecting a monitor<br>After adding or removing a monitor, the OS and window manager first have to detect this and will also decide on window placement. OpenBoard has to wait until this has settled before it tries to apply its own screen configuration. This timer is currently 3000ms.
- Screen configuration on startup or when otherwise showing a window for the firsttime. OpenBoard has to wait until the OS or window manager has applied its window placement policy before applying its own. This timer is currently 200ms.

If we stay with that timer-based approach then we have to choose the right values for these timers. There is always a tradeoff:

- If it is too short, then the desired effect is not achieved.
- If it is too long, then it might be a bad user experience.

Finally these timer values might be platform dependent. We could probably choose good initial values, large enough to cover all typical use cases, but make them configurable via settings.

### Board drawing area in multi-screen mode

When in multi-screen mode, the drawing area on the control screen has some unused space (to the left and right mainly), where also the background grid is not painted. Originally this should reflect how much is shown on the display screen, but with window mode it no longer works as expected. You now can resize the window, thus reducing the part of the drawing area visible on the control screen.

- Shouldn't we change this and always use the complete available drawing area on the control screen?
- Would it be useful to otherwise indicate the borders of the display screen on the control screen?
- Or is it just required to recalculate these margins when the window is resized? Currently they just stay the same and don't really make sense.

### Duplicate cursors

When I configure the pen to use the small circle as pointer then I observe the following in dual-monitor mode (display left of control, but I do not know whether it matters):

- When drawing on the board, all looks fine.
- When moving the mouse to the left on the display screen, then the crosshair cursor appears on the display screen. At the same time a second circle cursor appears on the control screen. However I cannot draw.

**Comment by kaamui** [comment]
> [fix: validation of screen configuration](https://github.com/OpenBoard-org/OpenBoard/commit/f8d5d69843206f00edd937766dd4a05349622c0f)
> 
> * never reject
> * make inacceptable input status visible by red border

Great improvements ! Just a couple of feedback : 

- the red and green borders can be hard to notice. For accessibility reasons, we should use a background as in Documents Mode (`editor->setStyleSheet("background-color: #FFB3C8;");`) and with the same color for consistency
- We should trim the values between every comma at the validation and when we persist the input, so something like "DP-1, DP-2" is an acceptable input
> * fixup list when screen label is pressed

Good idea !

> [fix: improve UBDisplayManager](https://github.com/OpenBoard-org/OpenBoard/commit/69e1b4024f55602b80e23c5c454639b97e8a772d)
> 
> * set initial window size in window mode
> * use the UBPlatformUtils.showFullScreen for control widget only
>   
>   * others must be unconditionally full screen
> * position screens after some delay
>   
>   * to override other decisions of the window manager
> * retain window size and position when switching screens in window mode

Windowed mode is working great now. The few feedback I can give concerning this specific part is the screens placement. The behavior is now as followed : 
- during changes applying, the issue with the Gnome Dock persists and `Display` screens are misplaced.
  1. exit -> relaunch, everything works as expected
  2. go to Desktop Mode, same behavior as when changed on-the-fly

The good news is that the issue itself is more predictable now. It seems that something is done at launch, making everything placed correctly, that is not done when switching to Desktop Mode or applying valid changes on the input. I don't understand why windows with a fullscreen flag should bother what's the available space. 
- When I switch off multiscreen button in the main menu, the input is not disabled 

- Focus issue is gone !

**Comment by kaamui** [comment]
> ### Explicitly apply screen configuration changes
> Currently changes of the screen configuration are applied on-the-fly as soon as a valid entry is provided in the screen list. There is the idea to replace this by explicit application of changes. I think there are some points to consider:
> 
> * All other configuration changes in the preferences dialog are applied on-the-fly, also the previous checkbox for switching control and display screen. Why should it be different for screen configuration?

My concern was that when deleting the entire input to change from "DP-1, DP-2, DP-3" to "DP-2, DP-3, DP-1" for example, changing what screen is the main one moves the input to the new main screen, and it can be quite disturbing to see (or even worse, not to see) this happen while your typing your new sequence. Note that you don't necessarily have all your screens right in front of you. 

That said, with all the improvements you made, I found it less disturbing during tests I made this mourning. 

> * Where should the "Apply" button be located? The first page of the preferences dialog is already quite full.
> 
on the right of the input ^^. Some parameters are probably going to be moved at some point this year in a tab dedicated to PDF options (where users will be able to choose to draw background color and/or grid at export, among other things ^^)

I'm not sure of what decision to make at this point so forget about it. I'll make some tries on my side when I'll start fully working  on 1.7

> ### Popup menu issue
> The following issue appearing on Windows has to be verified and analyzed:
> 
> > in FullScreen Mode, After going in Web Mode, main menu and context menus don't popup anymore. Returning to Board Mode, I open a textBox, right click on it and same issue. I think there is an issue with the window properties state (another clue : on my own PC at home, when entering Internal web browser, GeForce Experience overlay shows up like if I was entering in a game (fullscreen). I tried to record it to show you but recording stops when clicking on Web Mode button (maybe related to changes in the window properties ?)
> 

Still there. 

> 
> ### Use screen index instead of screen name in the UI
> The idea behind this point is to decouple from whatever screen names a specific operating system or window manager uses and to create a simple, not-surprising UI. While I support this point, there are still some questions open:
> 
> * How does the mapping to screen indexes work?
a vector of screen names (index in the vector is `index-1` that is displayed). Re-asked every time we need to display the screen indexes in the light rectangles ? Sorry if I missed the point that was worrying you.

> * Will the mapping change depending on which monitors are connected? If yes, is this then not again surprising to the user?

With how I think it would work, it would change. You could have entered or selected "1, 2, 3", close the preferences, remove the screen "2", open the preferences, and see "1,2" in the input. When clicking on the input, you'll understand why it changed. Some user will be definitely surprised though, but I don't think it should be this disturbing for most users. I didn't check so I could be wrong, but I presume it's what happens on Ubuntu if I unplug my screen "2" : the screen "3" won't be displayed as "3" anymore. 

> * What happens when a screen is connected or disconnected while the configuration dialog is open? Will we have to recalculate the mapping?

It would be great if we could, but I'm not sure we can detect if a screen has been removed or added while we're editing the input. But we could add a refresh button.

Note that I unplugged one of my screen, closed preferences then reopen it, and "DP-2-1" was now "DP-1" and "DP-1" was "DP-2"

> 
> ### Allow empty list
> If the list of screens is empty, this could be interpreted like the original multi-screen mode: Take and use all screens in the order as they are delivered from the OS. This would require that we distinguish between
> 
> * The configuration entry does not exist at all. This is the case when we upgrade from a previous version of OpenBoard.

For me this case should be treated as the second one. With no explicit value found in the configuration, we use the "standard" behavior : main scree only and input disabled in single screen mode, use every available screen available in multiscreen mode 
> * The configuration entry exists, but is empty. This is then the explicit configuration of the described multi-screen mode. To indicate this to the user we could then display a placeholder text in the line edit field stating "Use all connected screens".

Yes, it's a good idea, and no placeholder if in single screen mode.

**Comment by letsfindaway** [comment]
> Note that I unplugged one of my screen, closed preferences then reopened it, and "DP-2-1" was now "DP-1" and "DP-1" was "DP-2" (meaning that these names are not invariants)

This is an important finding I did not have in mind. I always thought (or better: hoped) that screen names are invariants and can be used to identify a video output, be it directly at the computer or at a docking station. As this assumption does not hold any longer I have to make some more considerations. What I want to achieve:

- Top priority: When the user connects the same monitors as before, the same screen configuration shall be applied.
- With another configuration of monitors, at least a reasonable configuration should be chosen.

This would in the end mean that we should store screen configurations (i.e. what shall be used by OpenBoard for what purpose) per monitor configuration (i.e. what monitors are connected). If a new monitor configuration is detected, then we do what we would do in case of an empty list (control screen on primary screen, display screen on next etc.). With your finding that even e.g. "DP-1" can get "DP-2" I would not try to reuse another configuration in that case.

A user could then configure screens for each monitor configuration. But this should be  hidden from the user behind the scenes. Implementation-wise this could be done with additional settings not managed by `UBSettings`, e.g.

```
ScreenList=DP-2,DP-1
ScreenList-DP-1,DP-2=DP-2,DP-1
ScreenList-DP-1,DP-2-1,DP-2-2=DP-1,DP-2-1
```

Only the first entry is actually configurable by the user. Behind the scenes this is synchronized to the proper entry below, depending on the current monitor configuration.

**Comment by kaamui** [comment]
I don't think it is a must (in the sense we need it to make things work), but it could be really cool indeed, if we obtain something really stable. Furthermore,  changing monitors or plug/unplug one is not happening every morning (expect for the QA team^^) so it should not generate a too important list in most common use cases. 

For me the top priority is that when deploying this implementation in thousands of classrooms (with heterogeneity in hardware), nothing has to be configured. The deployment team never had to configure monitors and everything has always been done automatically by OpenBoard. I think it will still be the case with your idea but I just wanted to write it so we can keep this in mind.

**Comment by letsfindaway** [comment]
> * When I switch off multiscreen button in the main menu, the input is not disabled

This intended. I don't think that selecting single monitor mode should disable the ability to configure a screen list. It just uses the first entry, but why should a user not be able to select the others, too?

**Comment by letsfindaway** [comment]
The current status of `improve-displaymanager` should now have a lot of stability fixes and I would like to make sure that the basic things actually work before looking into things like the board drawing area/boxing mode and related. I would be happy if we could check the list of topics you fund and see where we can put a checkmark!

And a question: The idea of showing a mouse cursor on the display screen in desktop mode or on a podcast sounds good to me. You might have seen my answer to the latest request, and in fact faking a mouse cursor are less than 10 lines of code.

My proposal is to add this feature at least to the API of the `UBDisplayManager`, by extending the `grab` function with an additional optional parameter `drawCursor` and then i a second step see where this could be useful, how it should be configurable etc. 

**Note:** for a mouse cursor on the display screen in desktop mode this is sufficient, for the podcast it needs more, as the podcast has various sources for the image - grabbing is just one - and we would need a faked cursor for each of them.

**Comment by kaamui** [comment]
> The current status of improve-displaymanager should now have a lot of stability fixes and I would like to make sure that the basic things actually work before looking into things like the board drawing area/boxing mode and related. I would be happy if we could check the list of topics you fund and see where we can put a checkmark!

Sadly, I'm going to work almost exclusively on something else than OpenBoard (I was supposed to start a month ago). It's going to be hard to be reactive on my side. For what I remember of my last testing session, it was behaving pretty well ! The only thing that is really blocking/critical is the issue on Windows (menus no longer working after entering in Web mode), but I need time to help you on all these issues and I don't have it, at least for now.  I'll try to do my best to look at it asap.

> And a question: The idea of showing a mouse cursor on the display screen in desktop mode or on a podcast sounds good to me. You might have seen my answer to the latest request, and in fact faking a mouse cursor are less than 10 lines of code.

Yep, I saw it. I also looked it a bit and saw some options in ffmpeg, but the difficulty is also that Windows doesn't work with ffmpeg, currently. I'm not 100% sure it should change though, as we have no feedback on hardware incompatibility, stability, etc despite we have some in ffmpeg-based implementations (macOS and linux).  

Your solution for grabbing has the advantage to be portable, and maybe it could be a nice way in the short-term future to handle it for recodring, providing the ability (in the podcast menu, where we choose audio, etc) for the user to grab the mouse cursor or not. If I understood/supposed well (but I read quickly, so I may say something stupid) you're just adding a QGraphicsItem looking like a mouse cursor at mouse cursor position before grabbing ?

At this point, too much things prevent me from making any decision about recording, grabbing, mirroring... instinctively, I think that there is probably something better to do than the current implementation of mirroring, but the difficulty is (I imagine) with capturing Desktop and other apps. A "streaming-oriented" solution must exist, but again, no time. Grabbing seems (to me) correct, especially with your recent work on it. Recording is hard to judge for me as I didn't implement it (just made ffmpeg implementation compatible with macOS) and considering that :  
- a recent survey showed that podcast is not a priority at all for our users
- I've a lack of low-level understanding of how ffmpeg or Win32 API/Windows Media SDK work
- I've no time  (I know, I should stop saying it, it would save me some :D)

**Comment by kaamui** [comment]
> [ ] Popup menu issue

It's not only related to Web Mode. By enabling startupHints (`EnableStartupHints=true` in OpenBoard.config, I was able to reproduce it  at start. I can't open the main menu or any other context menu

**Comment by letsfindaway** [comment]
merged

**Comment by kaamui** [comment]
A new issue appeared on Windows. A click on the line edit makes the app crash.

Tested on Qt 5.15.0 and Qt 6.4.0. Don't know where it comes from, as I remember well having tested this on Windows with no crash.

**Comment by kaamui** [comment]
Related to discussion on commit bfb88d9 but I put it here to not forget it. 

I found that screens attached to widgets are not changed until a certain point (internally by Qt), occurring after our own process (`adjustScreens`). I wonder if it could somehow be related to the cases where one oepnboard's display would no go on the desired screen. In Qt 6.4, QWidget::setScreen has been added. I tried to change it during positionScreens, before setGeometry and showFullScreen is called, without success.
